### PR TITLE
Updated to python3

### LINKED ISSUE #30: Update hepdata-converter to work with python3
Should build on #29

### PR DESCRIPTION
**Pull request by alisonrclarke**
Fixes #30

### CODE REVIEW

File: hepdata_converter/parsers/oldhepdata_parser.py [hunk 1]
```diff
@@ -126,7 +127,7 @@ def parse(self, data_in):
         # clean any possible data from previous parsing
         self.reset()
         # in case of strings we should treat them as filepaths
-        if isinstance(data_in, (str, unicode)):
+        if isinstance(data_in, (str, str)):
```
Comment (GraemeWatt) [inline hepdata_converter/parsers/oldhepdata_parser.py:130]: `if isinstance(data_in, str):`

File: hepdata_converter/testsuite/__init__.py [hunk 2]
```diff
@@ -98,16 +101,16 @@ class ExtendedTestCase(unittest.TestCase):
     def assertMultiLineAlmostEqual(self, first, second, msg=None):
         if hasattr(first, 'readlines'):
             lines = first.readlines()
-        elif isinstance(first, (str, unicode)):
+        elif isinstance(first, (str, str)):
```
Comment (GraemeWatt) [inline hepdata_converter/testsuite/__init__.py:104]: `elif isinstance(first, str):`

File: hepdata_converter/testsuite/__init__.py [hunk 3]
```diff
@@ -98,16 +101,16 @@ class ExtendedTestCase(unittest.TestCase):
...
             lines = first.split('\n')
 
         if hasattr(second, 'readlines'):
             orig_lines = second.readlines()
-        elif isinstance(second, (str, unicode)):
+        elif isinstance(second, (str, str)):
```
Comment (GraemeWatt) [inline hepdata_converter/testsuite/__init__.py:109]: `elif isinstance(second, str):`

File: hepdata_converter/writers/array_writer.py [hunk 4]
```diff
@@ -21,7 +24,7 @@ def is_number_var(cls, *variables):
         is_number_list = []
         for variable in variables:
             for element in variable['values']:
-                if 'value' in element and isinstance(element['value'], (str, unicode)):
+                if 'value' in element and isinstance(element['value'], (str, str)):
```
Comment (GraemeWatt) [inline hepdata_converter/writers/array_writer.py:27]: `if 'value' in element and isinstance(element['value'], str):`

File: hepdata_converter/writers/array_writer.py [hunk 5]
```diff
@@ -265,7 +266,7 @@ def _prepare_outputs(self, data_out, outputs):
                 for table in self.tables:
                     outputs.append(open(os.path.join(data_out, table.name.replace(' ','').replace('/','-').replace('$','').replace('\\','') + '.' + self.extension), 'w'))
         # multiple tables - require directory
-        elif len(self.tables) > 1 and not isinstance(data_out, (str, unicode)):
+        elif len(self.tables) > 1 and not isinstance(data_out, (str, str)):
```
Comment (GraemeWatt) [inline hepdata_converter/writers/array_writer.py:269]: `elif len(self.tables) > 1 and not isinstance(data_out, str):`

File: hepdata_converter/writers/array_writer.py [hunk 6]
```diff
@@ -367,19 +368,19 @@ def _parse_dependent_variable(cls, dependent_variable, headers, qualifiers, qual
...
                     has_error = False
-                    for i in xrange(len(value.get('errors', []))):
+                    for i in range(len(value.get('errors', []))):
                         error = value['errors'][i]
                         label = error.get('label', 'error')
 
                         if 'symerror' in error:
                             error_plus = error['symerror']
-                            if isinstance(error_plus, (str, unicode)):
+                            if isinstance(error_plus, (str, str)):
```
Comment (GraemeWatt) [inline hepdata_converter/writers/array_writer.py:383]: `if isinstance(error_plus, str):`

File: hepdata_converter/writers/root_writer.py [hunk 7]
```diff
@@ -419,7 +421,7 @@ def _prepare_outputs(self, data_out, outputs):
         :return:
         """
         compress = ROOTModule.ROOT.CompressionSettings(ROOTModule.ROOT.kZLIB, 1)
-        if isinstance(data_out, (str, unicode)):
+        if isinstance(data_out, (str, str)):
```
Comment (GraemeWatt) [inline hepdata_converter/writers/root_writer.py:424]: `if isinstance(data_out, str):`

File: hepdata_converter/writers/utils.py [hunk 8]
```diff
@@ -8,11 +10,11 @@ def error_value_processor(value, error):
     :param error: e.g. 20.0%
     :return: the absolute error, e.g. 12 for the above case.
     """
-    if isinstance(error, (str, unicode)):
+    if isinstance(error, (str, str)):
```
Comment (GraemeWatt) [inline hepdata_converter/writers/utils.py:13]: `if isinstance(error, str):`

File: hepdata_converter/writers/yaml_writer.py [hunk 9]
```diff
@@ -48,7 +49,7 @@ def write(self, data_in, data_out, *args, **kwargs):
             for table in tables:
                 table.metadata['table_doi'] = self.hepdata_doi + '/t' + str(table.index)
 
-        if not isinstance(data_out, (str, unicode)) and not self.single_file:
+        if not isinstance(data_out, (str, str)) and not self.single_file:
```
Comment (GraemeWatt) [inline hepdata_converter/writers/yaml_writer.py:52]: `if not isinstance(data_out, str) and not self.single_file:`

File: hepdata_converter/writers/yaml_writer.py [hunk 10]
```diff
@@ -60,7 +61,7 @@ def write(self, data_in, data_out, *args, **kwargs):
                     with open(os.path.join(data_out, table.data_file), 'w') as table_file:
                         yaml.dump(table.data, table_file, Dumper=Dumper, default_flow_style=None)
         else:
-            if isinstance(data_out, (str, unicode)):
+            if isinstance(data_out, (str, str)):
```
Comment (GraemeWatt) [inline hepdata_converter/writers/yaml_writer.py:64]: `if isinstance(data_out, str):`

File: hepdata_converter/writers/yoda_writer.py [hunk 11]
```diff
@@ -92,7 +94,7 @@ def __init__(self, *args, **kwargs):
         self.extension = 'yoda'
 
     def _prepare_outputs(self, data_out, outputs):
-        if isinstance(data_out, (str, unicode)):
+        if isinstance(data_out, (str, str)):
```
Comment (GraemeWatt) [inline hepdata_converter/writers/yoda_writer.py:97]: `if isinstance(data_out, str):`

File: hepdata_converter/writers/yoda_writer.py [hunk 12]
```diff
@@ -120,10 +122,8 @@ def _write_table(self, data_out, table):
                     if qualifier['name'] == 'Custom Rivet identifier':
                         rivet_identifier = qualifier['value']
             rivet_path = '/REF/' + self.rivet_analysis_name + '/' + rivet_identifier
-            graph.title = table_doi  # use for YODA 1.7.7
-            graph.path = rivet_path  # use for YODA 1.7.7
-#            graph.setTitle(table_doi)  # use for YODA 1.8.1
-#            graph.setPath(rivet_path)  # use for YODA 1.8.1
+            graph.setTitle(table_doi)  # use for YODA 1.8.1
+            graph.setPath(rivet_path)  # use for YODA 1.8.1
```
Comment (GraemeWatt) [inline hepdata_converter/writers/yoda_writer.py:126]: Change comment to `# use for YODA > 1.8.0`.

File: hepdata_converter/writers/utils.py [hunk 13]
```diff
@@ -8,11 +10,11 @@ def error_value_processor(value, error):
...
         try:
             if "%" in error:
                 error_float = float(error.replace("%", ""))
-                error_abs = (value/100) * error_float
+                error_abs = (old_div(value,100)) * error_float
```
Comment (GraemeWatt) [inline hepdata_converter/writers/utils.py:17]: Previous line was fine in Python 3, i.e. `(value/100) * error_float`.

File: hepdata_converter/writers/yoda_writer.py (hunk 12 above)
Comment (alisonrclarke) [reply hepdata_converter/writers/yoda_writer.py:124]: Should I just delete the comments now we don't have two versions of the code?

**Comment by alisonrclarke** [comment]
> The supported Python version (3.6 or later) should also be updated in `travis.yml`, `.readthedocs.yml`, and `setup.py`.

I'll change to python 3.6 in `.travis.yml` for consistency but it's worth noting that all the python code in the travis script is run via docker, so it shouldn't affect anything. On that point, do we need the install steps at https://github.com/DurhamARC/hepdata-converter/blob/e173022edbfe521726212c5059662c1e6975b16b/.travis.yml#L18 ?

**Comment by GraemeWatt** [comment]
> I'll change to python 3.6 in `.travis.yml` for consistency but it's worth noting that all the python code in the travis script is run via docker, so it shouldn't affect anything. On that point, do we need the install steps at https://github.com/DurhamARC/hepdata-converter/blob/e173022edbfe521726212c5059662c1e6975b16b/.travis.yml#L18 ?

Probably Python 3.7 is better to match the version used in the `hepdata-converter` Docker image.  I think it's only needed for running `coveralls`, i.e. push the `coverage` output to [coveralls.io](https://coveralls.io/github/HEPData/hepdata-converter?branch=master).  Probably the line `sudo pip install coverage` can be removed since `coverage` will be installed by `coveralls`.  It might not be ideal that `coverage` runs inside the Docker container, but `coveralls` runs outside, but it seems to work and I don't know if `coveralls` would work inside the Docker container.

File: hepdata_converter/writers/yoda_writer.py (hunk 12 above)
Comment (GraemeWatt) [reply hepdata_converter/writers/yoda_writer.py:124]: Yes, better just to delete the comment, since we won't downgrade YODA in future.

**Comment by alisonrclarke** [comment]
I think I've addressed all the comments now, but before this can be merged we need to ensure hepdata/hepdata-converter-docker#6 is OK, build a new docker image, and update both the docker image tag and the docker submodule here.

**Comment by GraemeWatt** [comment]
I've merged HEPData/hepdata-converter-docker/pull/6 and the updated `hepdata/hepdata-converter:0.2.0` image is available on [Docker Hub](https://hub.docker.com/r/hepdata/hepdata-converter).  Could you please update the `docker` submodule before I merge?

**Comment by GraemeWatt** [comment]
I had to restart the [Travis CI job](https://travis-ci.org/github/HEPData/hepdata-converter/builds/703218183) since there were some temporary problems this afternoon.  From the job log, it looks like `sudo pip install` uses Python 2.7 to install packages globally rather than installing for Python 3.7 in a virtual environment.  This causes problems at the end of the job when running `coveralls`.  Could you please remove the `sudo` from the `pip install` lines and also the deprecated `sudo: required` line from the `.travis.yml` file?

**Pull request merged**